### PR TITLE
feat: generate unlock sudt tx

### PR DESCRIPTION
**Pull request by pygman**

### CODE REVIEW

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 1]
```diff
@@ -802,6 +802,223 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+          capacity: '0x0',
+          lock: parseAddress(record.recipient),
+          type: sudtScript,
+        },
+        data: utils.toBigUInt128LE(amount),
+      };
+      const recipientCellMinimalCapacity = minimalCellCapacity(recipientCell);
+      totalSudtOutputCapacity += recipientCellMinimalCapacity;
+      totalSudtOutputAmount += amount;
+      recipientCell.cell_output.capacity = `0x${recipientCellMinimalCapacity.toString(16)}`;
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:869]: recipientCell capacity is a fixed number from config, ref:

https://github.com/nervosnetwork/force-bridge/blob/85725f19cae4f5c36eacf8e351e85568ac56c6fa/offchain-modules/packages/x/src/ckb/tx-helper/generator.ts#L247

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 2]
```diff
@@ -802,6 +802,223 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+    });
+    const searchedMultisigCellsCapacity = searchedMultisigCells
+      .map((cell) => BigInt(cell.cell_output.capacity))
+      .reduce((a, b) => a + b, 0n);
+
+    // add multisigChangeCell if needed
+    let omniLockMultisigChangeCellMinimalCapacity = BigInt(0);
+    if (searchedMultisigCellsAmount > totalSudtOutputAmount) {
+      const multisigChangeCellAmount = searchedMultisigCellsAmount - totalSudtOutputAmount;
+      omniLockMultisigChangeCell.data = `0x${multisigChangeCellAmount.toString(16)}`;
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:915]: use `utils.toBigUInt128LE`

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 3]
```diff
@@ -802,6 +802,223 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+      omniLockMultisigChangeCellMinimalCapacity = minimalCellCapacity(omniLockMultisigChangeCell);
+      omniLockMultisigChangeCell.cell_output.capacity = `0x${omniLockMultisigChangeCellMinimalCapacity.toString(16)}`;
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:917]: omniLockMultisigChangeCell.cell_output.capacity = multisig_cell_1_capacity_in_input +  multisig_cell_2_capacity_in_input + ...;

In other words, we recycle all the capacity from input multsigCells, and we use collector account to supply capacity to recipient cells and to tx fee.

```
if (searchedMultisigCellsAmount > totalSudtOutputAmount) {
  // all the capacity from input multsigCells are recycled in omniLockMultisigSudtChangeCell
}

if (searchedMultisigCellsAmount === totalSudtOutputAmount) {
  // all the capacity from input multsigCells are recycled in omniLockMultisigCKBChangeCell
}
```

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 4]
```diff
@@ -802,6 +802,223 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+    const collectorWitness = new Reader(
+      SerializeWitnessArgs(
+        normalizers.NormalizeWitnessArgs({
+          lock: SECP_SIGNATURE_PLACEHOLDER,
+        }),
+      ),
+    ).serializeJson();
+    txSkeleton = txSkeleton.update('witnesses', (witnesses) => {
+      return witnesses.push(collectorWitness);
+    });
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:976]: please fix this refer to https://github.com/nervosnetwork/force-bridge/pull/285

Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: fixed

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts (hunk 3 above)
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: if searchedMultisigCellsAmount === totalSudtOutputAmount then use ckb cell for change

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts (hunk 2 above)
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: fixed

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts (hunk 1 above)
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: Useful knowledge acquired

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 5]
```diff
@@ -802,6 +802,230 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+    const omniLockMultisigScript = parseAddress(getOmniLockMultisigAddress());
+    const searchKey = {
+      script: omniLockMultisigScript,
+      script_type: ScriptType.lock,
+      filter: {
+        script: sudtScript,
+      },
+    };
+
+    const searchedMultisigCells = await this.collector.collectSudtByAmountAndCapacity(
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:885]: tx struct is incorrect, please read design_doc/example_for_unlockSudt

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts [hunk 6]
```diff
@@ -802,6 +802,230 @@ export class CkbTxGenerator extends CkbTxHelper {
...
+      searchedMultisigCellsCapacity - totalSudtOutputCapacity
+    ).toString(16)}`;
+    txSkeleton = txSkeleton.update('inputs', (inputs) => {
+      return inputs.push(...searchedMultisigCells);
+    });
+    txSkeleton = txSkeleton.update('outputs', (outputs) => {
+      return outputs.push(omniLockMultisigChangeCell);
+    });
+
+    // add collector cell to pay tx fee
```
Comment (mkxbl) [inline offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:931]: use collector account to pay tx fee ***and supply capacity for recipient_sudt_cells***, that is to say, do not supply capacity for recipient_sudt_cells from committeeMultisigCells. ref: design_doc/example_for_unlockSudt

Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts:924]: unlock sudt capacity change and amount change calculation logic refactored according to design_doc/example_for_unlockSudt

File: offchain-modules/packages/x/src/ckb/tx-helper/generator.ts (hunk 5 above)
Comment (pygman) [reply offchain-modules/packages/x/src/ckb/tx-helper/generator.ts]: unlock sudt capacity change and amount change calculation logic refactored according to design_doc/example_for_unlockSudt

**Review by mkxbl — APPROVED** [review]

**Pull request merged**